### PR TITLE
Add snapcast

### PR DESCRIPTION
**Pull request by Saiyato**
This PR introduces SnapCast functionality to Volumio, there's still some fine tuning to do in terms of service support, I will elaborate below.

The plugin will install SnapCast for rpi devices (armhf), x86 is dependent on my testmachine, which has been failing recently. But it shouldn't require more than a change in the install script (add x86 arch and the corresponding SnapCast file).

After installation the plugin itself will write the configured options to various files.

### Snapserver/-client settings
Writes the supported settings/values to: /etc/default/snap{client|server}.

### MPD settings 
Updates /etc/mpd.conf, here you can en- or disable FIFO (stream) and ALSA (direct) output.

### Patch ALSA configuration
Updates asound.conf with the virtual device(s) needed for other services, it should support volspotconnect2, volspotconnect(1) was recently updates and needs rework.

### Spotify integration settings
Updates the Spotify implementation:

- For volspotconnect it will update the internal setting with the correct implementation (see next option)
- If dedicated stream is enabled, different options can apply (samplerate, bitdepth, channels), otherwise the standard stream (as configured in snapserver) is used.
- If librespot is used it will automatically create a new stream; read the SnapCast documentation for specifics.

### Patch Spotify configuration templates
Based on the above saved implementation, the templates in the corresponding plugin directories will be patched to enable multiroom playback (virtual soundcard output, piped to the fifo).

### Configure SnapServer streams
This enables you to change streams, usually done in the Android app, which allows for easier configuration. It also allows you to map the volume control to an instance.

### Current Volumio sound config
Only _shows_ the current output device and mixer, as configured in Volumio.

REWORK still to do:

- Patch the volspotconnect(1) implementation
- Read into metadata transfer, should be available
- Enable Airplay support
- Enable DLNA support

Updates in last commit:
- Removed hardcoded use of default password
- Removed node modules

### CODE REVIEW

**Comment by rayrrr** [comment]
@Saiyato I don't have write access to this repo but would like to see this merged soon...if you don't mind, please add some more details/context to this PR's description to aid review. Thanks!

**Comment by Saiyato** [comment]
This PR introduces SnapCast functionality to Volumio, there's still some fine tuning to do in terms of service support, I will elaborate below.

The plugin will install SnapCast for rpi devices (armhf), x86 is dependent on my testmachine, which has been failing recently. But it shouldn't require more than a change in the install script (add x86 arch and the corresponding SnapCast file).

After installation the plugin itself will write the configured options to various files.

### Snapserver/-client settings
Writes the supported settings/values to: /etc/default/snap{client|server}.

### MPD settings 
Updates /etc/mpd.conf, here you can en- or disable FIFO (stream) and ALSA (direct) output.

### Patch ALSA configuration
Updates asound.conf with the virtual device(s) needed for other services, it should support volspotconnect2, volspotconnect(1) was recently updates and needs rework.

### Spotify integration settings
Updates the Spotify implementation:

- For volspotconnect it will update the internal setting with the correct implementation (see next option)
- If dedicated stream is enabled, different options can apply (samplerate, bitdepth, channels), otherwise the standard stream (as configured in snapserver) is used.
- If librespot is used it will automatically create a new stream; read the SnapCast documentation for specifics.

### Patch Spotify configuration templates
Based on the above saved implementation, the templates in the corresponding plugin directories will be patched to enable multiroom playback (virtual soundcard output, piped to the fifo).

### Configure SnapServer streams
This enables you to change streams, usually done in the Android app, which allows for easier configuration. It also allows you to map the volume control to an instance.

### Current Volumio sound config
Only _shows_ the current output device and mixer, as configured in Volumio.

REWORK still to do:

- Patch the volspotconnect(1) implementation
- Read into metadata transfer, should be available
- Enable Airplay support
- Enable DLNA support

**Comment by rayrrr** [comment]
@Saiyato thanks for adding the detailed description. One tip: you added it as a PR _comment_ instead of putting it in the actual PR _description_ area at the top of the page where it currently says

> * Removed hardcoded use of default password
> * Removed node modules

Can you move your detailed description to there? I'm hoping it'll get more eyes at the top.

**Comment by Saiyato** [comment]
Done :)

**Comment by rayrrr** [comment]
Thanks @Saiyato! Now for bonus points, see if you can copy & paste it again from the comment to the description, _in a way that maintains the Markdown formatting._

**Comment by Saiyato** [comment]
Oh, haha, hadn't noticed... mea culpa, will fix later on ;)

File: plugins/miscellanea/snapcast/package.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	"net": "^1.0.2",
+	"object-component": "^0.0.3"
+  },
+  "directories": {
+    "example": "examples"
+  },
+  "devDependencies": {},
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Saiyato/volumio-snapcast-plugin.git"
```
Comment (rayrrr) [inline plugins/miscellanea/snapcast/package.json]: Update to reflect the volumio repo URL rather than your personal repo's

File: plugins/miscellanea/snapcast/package.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  },
+  "keywords": [
+    "snapcast",
+    "multiroom",
+    "audio",
+    "volumio"
+  ],
+  "bugs": {
+    "url": "https://github.com/Saiyato/volumio-snapcast-plugin/issues"
```
Comment (rayrrr) [inline plugins/miscellanea/snapcast/package.json]: Update to reflect the volumio repo URL rather than your personal repo's

File: plugins/miscellanea/snapcast/package.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  },
+  "homepage": "https://github.com/Saiyato/volumio-snapcast-plugin#readme"
```
Comment (rayrrr) [inline plugins/miscellanea/snapcast/package.json]: Update to reflect the volumio repo URL rather than your personal repo's

File: plugins/miscellanea/snapcast/install.sh [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+INSTALLING="/home/volumio/snapcast-plugin.installing"
+
+if [ ! -f $INSTALLING ]; then
+
+	touch $INSTALLING
+	apt-get update
+
+	# Download latest SnapCast packages
+	mkdir /home/volumio/snapcast
+	wget $(curl -s https://api.github.com/repos/badaix/snapcast/releases/latest | grep 'armhf' | cut -d\" -f4) -P /home/volumio/snapcast
```
Comment (rayrrr) [inline plugins/miscellanea/snapcast/install.sh]: Here is where we need to download a different file for x86 arch, correct? Is the idea to detect the system architecture here and then download the matching file for either rpi (armhf) or x86? And where else do we need to make changes for x86 exactly?

File: plugins/miscellanea/snapcast/default/snapclient [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+#   -p, --port arg (=1704)          server port
+#   -d, --daemon [=arg(=-3)]        daemonize, optional process priority [-20..19]
+#   --user arg                      the user[:group] to run snapclient as when daemonized
+#   --latency arg (=0)              latency of the soundcard
+#   -i, --instance arg (=1)         instance id
+#   --hostID arg                    unique host id
+
+USER_OPTS="--user snapclient:audio"
+
+SNAPCLIENT_OPTS="-d -h 127.0.0.1 -s sndrpihifiberry"
```
Comment (rayrrr) [inline plugins/miscellanea/snapcast/default/snapclient]: Can we replace `sndrpihifiberry` with `sndsnapclient` or something else not tied to a specific DAC vendor?

**Comment by balbuze** [comment]
What can we do with this?

**Comment by Saiyato** [comment]
Well, it should work on Pi's, have not yet had the time to fully test on x86. Let me see what I can do, I believe I can do the suggested changes by @rayrrr to make the plugin more generic. After that we should be good for ARM I believe.

**Comment by OneB1t** [comment]
root@volumio:/home/volumio/volumio-snapcast-plugin-master# volumio plugin install

This command will install the plugin on your device

Compressing the plugin
Plugin succesfully compressed
Progress: 10
Status :Downloading plugin

its stucked like that

**Comment by Saiyato** [comment]
Just made the URL's Volumio specific, still need to test x86 though.

File: plugins/miscellanea/snapcast/install.sh (hunk 4 above)
Comment (rayrrr) [reply plugins/miscellanea/snapcast/install.sh]: @Saiyato I have no idea, sorry.  The only way I can help here is with the spare x86 laptop sitting right on my desk. Happy to test if your old laptop is bricked.

**Comment by rayrrr** [comment]
I can help test x86 but have never built Volumio from source before so could use a bit of guidance.

**Comment by gkkpch** [comment]
@rayrrr: i've done that a few times ;) so I can help you with that. eMail me at volumio(at)bluewin.ch and I'll start you off with a small guide. 
To begin with, you need a debian jessie setup. A Virtual Box will also do, as long as it is running jessie.
Look at the README in the volumio/build repo for the prerequisites, don't forget to add the lzop package.

**Comment by Saiyato** [comment]
The biggest problem with x86 is that you need to build snapcast; or someone needs to pre-build the client and server and provide me with the deb packages. The make command will just create the binary and install it, but that also means you need to install a lot of other packages, just to build snapcast; which seems a bit over the top to me.

TL;DR: we need someone with build knowledge to build the x86 deb packages we can use to install. I'm am by no means knowledgeable enough at this point to do so. I tried and failed in various occasions to create a package which can be used on a fresh install.

**Comment by gkkpch** [comment]
yes, noticed the issue. Michelangelo did compile/ supply packages for arm, but they were not compiled for  i386. I'm not skilled enough yet to do .deb packaging, sorry can't help there.

**Comment by OneB1t** [comment]
> > root@volumio:/home/volumio/volumio-snapcast-plugin-master# volumio plugin install
> > This command will install the plugin on your device
> > Compressing the plugin
> > Plugin succesfully compressed
> > Progress: 10
> > Status :Downloading plugin
> > its stucked like that
> 
> Hi there,
> 
> I have no idea what you are trying to achieve, I'm going to need more info or you can wait till it's uploaded in the store, then you won't need to work around it anymore. ;)

I wanted to install snapcast with this command to arm based orange pi pc plus and it failed in state Downloading plugin :-/

**Comment by Saiyato** [comment]
> root@volumio:/home/volumio/volumio-snapcast-plugin-master# volumio plugin install
> 
> This command will install the plugin on your device
> 
> Compressing the plugin
> Plugin succesfully compressed
> Progress: 10
> Status :Downloading plugin
> 
> its stucked like that

What does echo `$(lscpu | awk 'FNR == 1 {print $2}')` give you? Plus, can you download from github from the machine? And can you post the systemd journal entries from around that time? That would help tremendously. 

Also, how did you download the plugin and did you install the necessary node_modules prior to installing? Not entirely sure if `volumio plugin install` does that for you already.

If you are going to install from github you will need some knowledge about the system, there are many variables on which installation can fail.

**Comment by Saiyato** [comment]
Update, just installed on not-supported architecture:

```
This command will install the plugin on your device

Compressing the plugin
Plugin succesfully compressed
Progress: 10
Status :Downloading plugin
Progress: 30
Status :Creating folder on disk
Progress: 40
Status :Unpacking plugin
Progress: 50
Status :Checking for duplicate plugin
Progress: 60
Status :Copying Plugin into location
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 70
Status :Installing dependencies
Progress: 90
Status :Adding plugin to registry
Progress: 100
Status :Plugin Successfully Installed
Done!
Plugin installed successfully!
```

The matching journal entries are:
```
Apr 25 08:44:24 volumiox86 volumio[942]: info: Downloading plugin at http://127.0.0.1:3000/plugin-serve/snapcast.zip
Apr 25 08:44:24 volumiox86 volumio[942]: info: END DOWNLOAD: http://127.0.0.1:3000/plugin-serve/snapcast.zip
Apr 25 08:44:24 volumiox86 volumio[942]: info: Folder /tmp/plugins removed
Apr 25 08:44:25 volumiox86 volumio[942]: info: Checking if plugin already exists
Apr 25 08:44:25 volumiox86 volumio[942]: info: Rename folder
Apr 25 08:44:25 volumiox86 volumio[942]: info: Folder /tmp/downloaded_plugin.zip removed
Apr 25 08:44:25 volumiox86 volumio[942]: info: Move to category
Apr 25 08:44:29 volumiox86 volumio[942]: info: Checking if install.sh is present
Apr 25 08:44:29 volumiox86 volumio[942]: info: Executing install.sh
Apr 25 08:44:29 volumiox86 sudo[31777]: volumio : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/sh /data/plugins/miscellanea/snapcast/install.sh
Apr 25 08:44:29 volumiox86 sudo[31777]: pam_unix(sudo:session): session opened for user root by (uid=0)
Apr 25 08:44:29 volumiox86 volumio[942]: info: Installing SnapCast and its dependencies...
Apr 25 08:44:29 volumiox86 volumio[942]: info: Ign http://ftp.nl.debian.org jessie InRelease
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie Release.gpg
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie Release
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/main Sources
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/contrib Sources
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/non-free Sources
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/main i386 Packages
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/contrib i386 Packages
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/non-free i386 Packages
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/contrib Translation-en
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/main Translation-en
Apr 25 08:44:29 volumiox86 volumio[942]: info: Hit http://ftp.nl.debian.org jessie/non-free Translation-en
Apr 25 08:45:11 volumiox86 volumio[942]: info: Reading package lists...
Apr 25 08:45:15 volumiox86 volumio[942]: info: Selecting previously unselected package snapclient.
Apr 25 08:45:15 volumiox86 volumio[942]: info: (Reading database ... 29328 files and directories currently installed.)
Apr 25 08:45:15 volumiox86 volumio[942]: info: Processing triggers for man-db (2.7.0.2-5) ...
Apr 25 08:45:20 volumiox86 volumio[942]: info: Reading package lists...
Apr 25 08:45:21 volumiox86 volumio[942]: info: Building dependency tree...
Apr 25 08:45:21 volumiox86 volumio[942]: info: Reading state information...
Apr 25 08:45:21 volumiox86 volumio[942]: info: 0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Apr 25 08:45:21 volumiox86 systemd[1]: Reloading.
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:18] Unknown lvalue 'ProtectKernelTunables' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:19] Unknown lvalue 'ProtectControlGroups' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:20] Unknown lvalue 'ProtectKernelModules' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:23] Unknown lvalue 'RestrictNamespaces' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: Starting Creates a fifo buffer for spotify-connect-web/librespot...
Apr 25 08:45:21 volumiox86 systemd[1]: Started Creates a fifo buffer for spotify-connect-web/librespot.
Apr 25 08:45:21 volumiox86 spotfifo[32400]: rm: cannot remove '/tmp/snapfifo': Operation not permitted
Apr 25 08:45:21 volumiox86 systemd[1]: spotififo.service: main process exited, code=exited, status=1/FAILURE
Apr 25 08:45:21 volumiox86 systemd[1]: Unit spotififo.service entered failed state.
Apr 25 08:45:21 volumiox86 systemd[1]: Reloading.
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:18] Unknown lvalue 'ProtectKernelTunables' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:19] Unknown lvalue 'ProtectControlGroups' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:20] Unknown lvalue 'ProtectKernelModules' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:23] Unknown lvalue 'RestrictNamespaces' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: Reloading.
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:18] Unknown lvalue 'ProtectKernelTunables' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:19] Unknown lvalue 'ProtectControlGroups' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:20] Unknown lvalue 'ProtectKernelModules' in section 'Service'
Apr 25 08:45:21 volumiox86 systemd[1]: [/lib/systemd/system/mpd.service:23] Unknown lvalue 'RestrictNamespaces' in section 'Service'
Apr 25 08:45:22 volumiox86 volumio[1749]: ALSA lib conf.c:1852:(snd_config_load1) _toplevel_:21:0:Unexpected char
Apr 25 08:45:22 volumiox86 volumio[1749]: ALSA lib conf.c:3615:(config_file_open) /etc/asound.conf may be old or corrupted: consider to remove or fix it
Apr 25 08:45:22 volumiox86 volumio[1749]: ALSA lib conf.c:3537:(snd_config_hooks_call) function snd_config_hook_load returned error: Invalid argument
Apr 25 08:45:22 volumiox86 volumio[1749]: ALSA lib conf.c:3986:(snd_config_update_r) hooks failed, removing configuration
Apr 25 08:45:22 volumiox86 volumio[1749]: amixer: Mixer attach hw:0 error: Invalid argument
Apr 25 08:45:22 volumiox86 systemd[1]: Stopping Music Player Daemon...
Apr 25 08:45:22 volumiox86 systemd[1]: Starting Music Player Daemon...
Apr 25 08:45:22 volumiox86 systemd[1]: Started Music Player Daemon.
Apr 25 08:45:22 volumiox86 sudo[31777]: pam_unix(sudo:session): session closed for user root
Apr 25 08:45:22 volumiox86 volumio[942]: Plugin install end detected on script
Apr 25 08:45:22 volumiox86 volumio[942]: info: Install script completed
Apr 25 08:45:22 volumiox86 volumio[942]: info: Adding reference to registry
Apr 25 08:45:22 volumiox86 volumio[942]: info: Done installing plugin.
Apr 25 08:45:22 volumiox86 volumio[942]: info: Folder /tmp/plugins removed
Apr 25 08:45:22 volumiox86 volumio[942]: info: Folder /tmp/downloaded_plugin.zip removed
Apr 25 08:45:22 volumiox86 volumio[942]: info: Folder /data/temp removed

```

**Comment by balbuze** [comment]
@Saiyato is it ready to be merged?

**Comment by Saiyato** [comment]
Hi @balbuze , just removed the x86/x64 steps because the compiled packages is not yet up to par. So yes you can, but there's no x86 support yet. Still working on that. ;)

**Comment by balbuze** [comment]
done!

**Comment by Saiyato** [comment]
Awesome, thanks!
